src/analysis/normed_space/normed_group_hom_completion.lean

```diff
@@ -0,0 +1,271 @@
+/-
+Copyright (c) 2021 Patrick Massot. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Patrick Massot
+-/
+
+import analysis.normed_space.normed_group_hom
+/-!
+# Completion of normed group homs
+
+Given two (semi) normed groups `G` and `H` and a normed group hom `f  : normed_group_hom G H`,
+we build and study a normed group hom
+`f.completion  : normed_group_hom (completion G) (completion H)` such that the diagram
+
+                   f
+     G       ----------->     H
+
+     |                        |
+     |                        |
+     |                        |
+     V                        V
+
+completion G -----------> completion H
+            f.completion
+
+commutes. The map itself comes the general theory of completion of uniform spaces, but here
+we want a normed group hom, study its operator norm and kernel.
+
+The vertical maps in the above diagrams are also normed group homs constructed in this file.
+
+## Main definitions and results:
+
+* `normed_group_hom.completion`: see the discussion above.
+* `normed_group.to_compl : normed_group_hom G (completion G)`: the canonical map from `G` to its
+  completion, as a normed group hom
+* `normed_group_hom.completion_to_compl`: the above diagram indeed commutes.
+* `normed_group_hom.norm_completion_le`: `∥f.completion∥ ≤ ∥f∥`
+* `normed_group_hom.ker_le_ker_completion`: the kernel of `f.completion` contains the image of the
+  kernel of `f`.
+* `normed_group_hom.ker_completion`: the kernel of `f.completion` is the closure of the image of the
+  kernel of `f` under an assumption that `f` is quantitatively surjective onto its image.
+-/
+
+noncomputable theory
+
+open set normed_group_hom uniform_space uniform_space.completion
+
+variables {G : Type*} [semi_normed_group G]
+variables {H : Type*} [semi_normed_group H]
+variables {K : Type*} [semi_normed_group K]
+
+/-- The normed group hom induced between completions. -/
+def normed_group_hom.completion (f : normed_group_hom G H) :
+  normed_group_hom (completion G) (completion H) :=
+{ to_fun := completion.map f,
+  map_add' := begin
```

Comment:
Is there not yet the completion of a continuous group hom between uniform groups? That should give this `map_add'` for free, I think.

Reply:
I'll work on bringing `topology/algebra/group_completion.lean` to bundled homs.